File: packages/core/src/detect.ts
/** Raw OBS config files, read by the host and parsed here (no I/O in this module). */
export interface ObsConfigFiles {
  /** Contents of `obs-studio/user.ini`, or null if absent. */
  userIni: string | null;
  /** Map of profile directory name → contents of its `basic.ini`. */
  profiles: Record<string, string>;
}

/**
 * Read a single `key=value` from an INI section. Section and key match
 * case-insensitively; tolerant of a UTF-8 BOM and CRLF line endings, both of
 * which OBS writes.
 */
function getIniValue(text: string, section: string, key: string): string | undefined {
  let current: string | undefined;
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.replace(/^﻿/, "").trim();
    if (!line || line.startsWith(";") || line.startsWith("#")) continue;
    const header = /^\[(.+)\]$/.exec(line);
    if (header) {
      current = header[1].toLowerCase();
      continue;
    }
    if (current !== section.toLowerCase()) continue;
    const eq = line.indexOf("=");
    if (eq < 0) continue;
    if (line.slice(0, eq).trim().toLowerCase() === key.toLowerCase()) {
      return line.slice(eq + 1).trim();
    }
  }
  return undefined;
}

/**
 * Resolve the folder OBS records into, from its config files.
 *
 * Picks the active profile (`user.ini` `[Basic] ProfileDir`, falling back to
 * `Profile`, then the sole/first profile present), then reads that profile's
 * `basic.ini`: `[Output] Mode = Advanced` uses `[AdvOut] RecFilePath`, otherwise
 * `[SimpleOutput] FilePath`. Returns `undefined` when nothing usable is found.
 */
export function detectObsRecordingFolder(files: ObsConfigFiles): string | undefined {
  const profileNames = Object.keys(files.profiles);
  if (profileNames.length === 0) return undefined;

  const active =
    (files.userIni && getIniValue(files.userIni, "Basic", "ProfileDir")) ||
    (files.userIni && getIniValue(files.userIni, "Basic", "Profile"));

  const basicIni =
    (active && files.profiles[active]) ??
    (active && matchProfileCaseInsensitive(files.profiles, active)) ??
    files.profiles[profileNames[0]];

  if (!basicIni) return undefined;

  const mode = getIniValue(basicIni, "Output", "Mode");
  const folder =
    mode?.toLowerCase() === "advanced"
      ? getIniValue(basicIni, "AdvOut", "RecFilePath")
      : getIniValue(basicIni, "SimpleOutput", "FilePath");

  const trimmed = folder?.trim();
  return trimmed ? trimmed : undefined;
}

function matchProfileCaseInsensitive(
  profiles: Record<string, string>,
  name: string,
): string | undefined {
  const lower = name.toLowerCase();
  for (const [key, value] of Object.entries(profiles)) {
    if (key.toLowerCase() === lower) return value;
  }
  return undefined;
}

File: desktop/src/Qlipq.Core/Detect.cs
using System.Text.RegularExpressions;

namespace Qlipq.Core;

/// <summary>Raw OBS config files, read by the host and parsed here (no I/O in this module).</summary>
public sealed record ObsConfigFiles
{
    /// <summary>Contents of <c>obs-studio/user.ini</c>, or null if absent.</summary>
    public string? UserIni { get; set; }

    /// <summary>Map of profile directory name → contents of its <c>basic.ini</c>.</summary>
    public Dictionary<string, string> Profiles { get; set; } = [];
}

public static partial class Detect
{
    [GeneratedRegex(@"\r?\n")]
    private static partial Regex LineSplitRegex();

    [GeneratedRegex(@"^\[(.+)\]$")]
    private static partial Regex SectionHeaderRegex();

    /// <summary>
    /// Read a single <c>key=value</c> from an INI section. Section and key match
    /// case-insensitively; tolerant of a UTF-8 BOM and CRLF line endings, both of
    /// which OBS writes.
    /// </summary>
    private static string? GetIniValue(string text, string section, string key)
    {
        string? current = null;
        foreach (var raw in LineSplitRegex().Split(text))
        {
            var line = (raw.StartsWith('﻿') ? raw[1..] : raw).Trim();
            if (line.Length == 0 || line.StartsWith(';') || line.StartsWith('#')) continue;
            var header = SectionHeaderRegex().Match(line);
            if (header.Success)
            {
                current = header.Groups[1].Value.ToLowerInvariant();
                continue;
            }
            if (current != section.ToLowerInvariant()) continue;
            var eq = line.IndexOf('=');
            if (eq < 0) continue;
            if (line[..eq].Trim().ToLowerInvariant() == key.ToLowerInvariant())
            {
                return line[(eq + 1)..].Trim();
            }
        }
        return null;
    }

    /// <summary>
    /// Resolve the folder OBS records into, from its config files.
    /// Picks the active profile (<c>user.ini</c> <c>[Basic] ProfileDir</c>, falling back to
    /// <c>Profile</c>, then the sole/first profile present), then reads that profile's
    /// <c>basic.ini</c>: <c>[Output] Mode = Advanced</c> uses <c>[AdvOut] RecFilePath</c>,
    /// otherwise <c>[SimpleOutput] FilePath</c>. Returns <c>null</c> when nothing usable is found.
    /// </summary>
    public static string? DetectObsRecordingFolder(ObsConfigFiles files)
    {
        var profileNames = files.Profiles.Keys.ToList();
        if (profileNames.Count == 0) return null;

        string? active = null;
        if (!string.IsNullOrEmpty(files.UserIni))
        {
            active = GetIniValue(files.UserIni, "Basic", "ProfileDir");
            if (string.IsNullOrEmpty(active)) active = GetIniValue(files.UserIni, "Basic", "Profile");
        }

        string? basicIni = null;
        if (!string.IsNullOrEmpty(active))
        {
            basicIni = files.Profiles.TryGetValue(active, out var exact)
                ? exact
                : MatchProfileCaseInsensitive(files.Profiles, active);
        }
        basicIni ??= files.Profiles[profileNames[0]];

        if (basicIni is null) return null;

        var mode = GetIniValue(basicIni, "Output", "Mode");
        var folder = string.Equals(mode, "advanced", StringComparison.OrdinalIgnoreCase)
            ? GetIniValue(basicIni, "AdvOut", "RecFilePath")
            : GetIniValue(basicIni, "SimpleOutput", "FilePath");

        var trimmed = folder?.Trim();
        return string.IsNullOrEmpty(trimmed) ? null : trimmed;
    }

    private static string? MatchProfileCaseInsensitive(Dictionary<string, string> profiles, string name)
    {
        foreach (var (key, value) in profiles)
        {
            if (string.Equals(key, name, StringComparison.OrdinalIgnoreCase)) return value;
        }
        return null;
    }
}
